### PR TITLE
fix knife ssh --exit-on-error

### PR DESCRIPTION
**Pull request by tchernomax**
With enhancement #5978 `knife ssh` hang on some occasions.

```
root@test:/# knife ssh 'name:test.wwwpp*' "echo 'toto'; true" … -C 1 --exit-on-error
test.wwwpp-01   toto
test.wwwpp-02   toto

→ OK

root@test:/# knife ssh 'name:test.wwwpp*' "echo 'toto'; false" … -C 1 --exit-on-error
test.www-01   toto

→ OK

root@test:/# knife ssh 'name:test.www*' "echo 'toto'; true" … -C 1 --exit-on-error
test.www-01   toto
test.www-02   toto
test.wwwpp-02 toto
test.wwwpp-01 toto

→ OK

root@test:/# knife ssh 'name:test.www*' "echo 'toto'; false" … -C 1 --exit-on-error
test.www-01   toto
^C

→ FAIL, `knife ssh` hang
```

This patch fix this issue :

```
root@test:/# knife ssh 'name:test.wwwpp*' "echo 'toto'; true" … -C 1 --exit-on-error
test.wwwpp-01   toto
test.wwwpp-02   toto

→ OK

root@test:/# knife ssh 'name:test.wwwpp*' "echo 'toto'; false" … -C 1 --exit-on-error
test.www-01   toto

→ OK

root@test:/# knife ssh 'name:test.www*' "echo 'toto'; true" … -C 1 --exit-on-error
test.www-01   toto
test.www-02   toto
test.wwwpp-02 toto
test.wwwpp-01 toto

→ OK

root@test:/# knife ssh 'name:test.www*' "echo 'toto'; false" … -C 1 --exit-on-error
test.www-01   toto

→ OK

```

With -C 2 :

```
root@test:/# knife ssh 'name:test.www* OR name:test.api*' "echo 'toto'; true" … -C 2 --exit-on-error
test.api-02   toto
test.apipp-02 toto
test.www-02   toto
test.api-04   toto
test.www-01   toto
test.api-01   toto
test.wwwpp-01 toto
test.apipp-01 toto
test.api-03   toto
test.wwwpp-02 toto

→ OK

root@test:/# knife ssh 'name:test.www* OR name:test.api*' "echo 'toto'; false" … -C 2 --exit-on-error
test.apipp-02 toto
test.api-02   toto

→ OK

```

### Description

`knife ssh … --exit-on-error` ; end sessions loop in case of error.

### Issues Resolved

* #5978
* #5399

### Check List

- [ ] New functionality includes tests
- [ ] All tests pass
- [ ] RELEASE\_NOTES.md, has been updated if required (not required for bugfixes, required for API changes)
- [ ] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>

### CODE REVIEW

**Comment by tchernomax** [comment]
I don't know if this fail https://travis-ci.org/chef/chef/jobs/222085515#L1823 is "normal" since I modified ssh_command (I guess the problem is in the [before](https://github.com/chef/chef/blob/master/spec/unit/knife/ssh_spec.rb#L239) block).

Can someone explain me this ssh_command unit test ?

File: lib/chef/knife/ssh.rb [hunk 1]
```diff
@@ -332,7 +332,9 @@ def ssh_command(command, subsession = nil)
         command.force_encoding("binary") if command.respond_to?(:force_encoding)
         subsession.open_channel do |chan|
           if config[:on_error] && exit_status != 0
-            chan.close()
+            chan.exec "true" do |ch, success|
+              raise ArgumentError, "Cannot execute #{command}" unless success
+            end
```
Comment (lamont-granquist) [inline lib/chef/knife/ssh.rb]: firing off a `true` command here and using the success or failure of that command feels very hacky and side-effecty and brittle.

File: lib/chef/knife/ssh.rb [hunk 2]
```diff
@@ -332,7 +332,9 @@ def ssh_command(command, subsession = nil)
         command.force_encoding("binary") if command.respond_to?(:force_encoding)
         subsession.open_channel do |chan|
           if config[:on_error] && exit_status != 0
-            chan.close()
```
Comment (lamont-granquist) [inline lib/chef/knife/ssh.rb]: and this feels like we're workaround around a bug in net-ssh / net-ssh-multi somewhere that should get (or already is?) fixed properly

**Comment by lamont-granquist** [comment]
nobody seems to be piping up that this still exists as a bug, and we haven't had the time to retest and clean this up...

**Pull request closed without merging**